src/xds/monitoring.rs

```diff
@@ -0,0 +1,50 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use prometheus_client::{
+    encoding::text::Encode, metrics::counter::Counter, metrics::family::Family, registry::Registry,
+};
+
+#[derive(Clone, Hash, PartialEq, Eq, Encode)]
+pub struct ConnectionTerminationLabel {
+    reason: ConnectionTerminationReason,
+}
+
+#[derive(Clone, Hash, PartialEq, Eq, Encode)]
+pub enum ConnectionTerminationReason {
+    ConnectionError,
+    Error,
+}
+
+#[derive(Default, Clone)]
+pub struct XdsMetrics {
+    connection_terminations: Family<ConnectionTerminationLabel, Counter>,
+}
+
+impl XdsMetrics {
+    pub fn register(&self, registry: &mut Registry) {
+        let sub_registry = registry.sub_registry_with_prefix("xds");
+        sub_registry.register(
+            "connection_terminations",
```

Comment:
Any metrics not already part of Istio API should be heavily scrutinized since we are effectively making an API. So we can keep this for now but should be cautious

Reply:
Is anyone working on API design doc?

Reply:
@jewertow might be, not sure

Reply:
I'm not working on API design. Steven suggested to look at Envoy [cluster stats](https://www.envoyproxy.io/docs/envoy/latest/configuration/upstream/cluster_manager/cluster_stats#config-cluster-manager-cluster-stats), so I'm following that metrics naming, i.e. `istio_downstream_cx`, `istio_downstream_rq`, `istio_upstream_` and so on. But these are just names, you can suggest any prefix. We can follow istio-proxy [metrics](https://istio.io/latest/docs/reference/config/metrics/#metrics) naming as well, i.e. `istio_requests_total`, `istio_tcp_sent_bytes_total`, etc.

Reply:
IMO:
* anything not in https://istio.io/latest/docs/reference/config/metrics/ constitutes a new API
  * I am going to explore an idea of "experimental" or "internal" metrics so we can go wild with whatever metrics we want.
  * if we don't have 'experimental' we need API review for every metric, IMO, which limits our ability to add helpful debugging info
* Envoy metrics should be used as a reasonable reference for the *types of things* we want metrics for, but not metrics names, which are inscrutable  
* We definitely should implement as much of https://istio.io/latest/docs/reference/config/metrics/ constitutes a new API as we can (ie the 4 TCP ones)